### PR TITLE
first draft of "interacting with devices" for review

### PR DESCRIPTION
**Pull request by pat-rogers**
(Note than "handling interrupts' is not ready for review.)

### CODE REVIEW

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst [hunk 1]
```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
...
+
+Let's talk more about the type we use to represent a memory-mapped
+device. As we said, that type defines the view we have for the object,
+and hence the operations we have available for accessing the underlying
+mapped device.
+
+We choose the type for the representative Ada variable based on the
+interface of the hardware mapped to the memory. If the interface is a
+single monolithic register, for example, then an integer, either signed
+or unsigned, and of the necessary size, will suffice. But suppose the
```
Comment (gusthoff) [inline content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:270]: This phrase sounds a bit confusing to me. (Maybe there are too many commas here.) As a suggestion, what about using an m-dash?

```suggestion
interface of the hardware mapped to the memory. If the interface is a
single monolithic register, for example, then an integer |mdash| either signed
or unsigned, and of the necessary size |mdash| will suffice. But suppose the
```

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst [hunk 2]
```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
...
+of the :ada:`AFR`:
+
+.. code-block:: ada
+
+   type Alternate_Function_Fields is array (GPIO_Pin) of GPIO_Alternate_Function_Code
+     with Component_Size => 4, Size => 64;  -- both in units of bits
+
+Note that we can use the GPIO :ada:`Pin` parameter directly as the index into
+the array type, obviating any need to massage the :ada:`Pin` value in
+the procedure. That's possible because the type :ada:`GPIO_Pin` is an
```
Comment (gusthoff) [inline content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:448]: ``GPIO_Pin`` isn't defined in any of the code examples. Maybe we should have a type declaration here?

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst [hunk 3]
```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
...
+
+.. code-block:: c
+
+   temp = ((uint32_t)(GPIO_AF) << ((uint32_t)((uint32_t)GPIO_PinSource & (uint32_t)0x07) * 4)) ;
+   GPIOx->AFR[GPIO_PinSource >> 0x03] &= ~((uint32_t)0xF << ((uint32_t)((uint32_t)GPIO_PinSource & (uint32_t)0x07) * 4)) ;
+   temp_2 = GPIOx->AFR[GPIO_PinSource >> 0x03] | temp;
+   GPIOx->AFR[GPIO_PinSource >> 0x03] = temp_2;
+
+That's unfair to ask, absent any context. The code configures a general
+purpose I/O (GPIO) pin on an Arm microcontroller for one of the
```
Comment (gusthoff) [inline content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:42]: Minor editorial question: isn't the uppercase abbreviation ("ARM") more commonly used than "Arm"?

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst [hunk 4]
```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
...
+features or capacity. The gaps are thus between consecutive hardware
+devices.
+
+These gaps are not necessarily included in the memory layout documented
+for the device. You should check, therefore, that the documented
+starting addresses of the second and subsequent array components are
+what you will get with a simple array object having components of that
+record type.
+
+To represent the gap, an "extra", unused record component can be added,
```
Comment (gusthoff) [inline content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:588]: A simple / short code example would be nice.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 1 above)
Comment (pat-rogers) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:270]: Or maybe make it parenthetical, like this:

If the interface is a single monolithic register, for example, then an integer (signed
or unsigned) of the necessary size will suffice.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 2 above)
Comment (pat-rogers) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:448]: Will do.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 3 above)
Comment (pat-rogers) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:42]: IIUC "Arm" is the newer form, as it is no longer an acronym. To be checked.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 4 above)
Comment (pat-rogers) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:588]: Will do

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 1 above)
Comment (gusthoff) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:270]: That sounds good to me.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst [hunk 5]
```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
...
+and stores. If we wrote this in Ada it would look like this:
+
+.. code-block:: ada
+
+   procedure Lock (Port : in out GPIO_Port;  Pin : GPIO_Pin) is
+      Temp : Word with Volatile;
+   begin
+      --  set the lock control bit and the pin bit, clear the others
+      Temp := LCCK or Pin'Enum_Rep;
+      --  write the lock and pin bits
```
Comment (gusthoff) [inline content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:1273]: Minor editorial suggestion: an empty line before each comment could improve readability a bit.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst [hunk 6]
```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
...
+      --  read the lock bit again
+      Temp := Port.LCKR;
+   end Lock;
+
+:ada:`Temp` is marked volatile for the sake of getting exactly the load
+and stores that we express in the source code, corresponding to the
+hardware locking protocol.
+
+This high-level coding approach will work, and is simple enough that not
+using MCIs is not unreasonable. However, what really argues against it
```
Comment (gusthoff) [inline content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:1290]: Minor comment... There are three negatives here:

> not using MCIs is not unreasonable.

Maybe we could rewrite this to:

> not using MCIs can be reasonable.

or something similar.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 5 above)
Comment (pat-rogers) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:1273]: Sometimes I agree with that, sometimes not so much. I think it is overdone in the AdaCore style, for example.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 6 above)
Comment (pat-rogers) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:1290]: How about:

This high-level coding approach will work, and is simple enough that MCIs need not be used.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 5 above)
Comment (gusthoff) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:1273]: I was actually just referring to the readability of this specific code example. Here, I think an empty line could look better.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 6 above)
Comment (gusthoff) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:1290]: Perfect. :-)

**Review by gusthoff — CHANGES_REQUESTED** [review]
I'll wait for the revised version...

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 3 above)
Comment (pat-rogers) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:42]: Confirmed "Arm" is the current casing.

File: content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst (hunk 5 above)
Comment (pat-rogers) [reply content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst:1273]: OK

**Review by gusthoff — APPROVED** [review]

**Pull request merged**